src/main/java/org/commcare/core/network/ModernHttpRequester.java

```diff
@@ -41,149 +40,93 @@
-        }
-    }
-
-    protected HttpURLConnection setupConnection(URL builtUrl) throws IOException {
-        HttpURLConnection httpConnection = (HttpURLConnection)builtUrl.openConnection();
-        setupConnectionInner(httpConnection);
-        if (isPostRequest) {
-            httpConnection.setRequestMethod("POST");
-            httpConnection.setDoOutput(true);
-            buildPostPayload(httpConnection);
-        } else {
-            httpConnection.setRequestMethod("GET");
+    /**
+     * Executes the HTTP Request. Can be called directly to bypass response processor.
+     *
+     * @return Response from the HTTP call
+     * @throws IOException
+     */
+    public Response<ResponseBody> makeRequest() throws IOException {
+        switch (method) {
+            case POST:
+                currentCall = commCareNetworkService.makePostRequest(url, params, addContentHeaders(requestBody), requestBody);
+                break;
+            case MULTIPART_POST:
+                currentCall = commCareNetworkService.makeMultipartPostRequest(url, params, headers, parts);
+                break;
+            case GET:
+                currentCall = commCareNetworkService.makeGetRequest(url, params, headers);
+                break;
+            default:
+                throw new IllegalArgumentException("Invalid HTTPMethod " + method.toString());
         }
-
-        return httpConnection;
-    }
-
-    private static void setupConnectionInner(HttpURLConnection httpConnection) {
-        httpConnection.setConnectTimeout(CONNECTION_TIMEOUT);
-        httpConnection.setReadTimeout(CONNECTION_SO_TIMEOUT);
-        httpConnection.setDoInput(true);
-        httpConnection.setInstanceFollowRedirects(true);
+        return currentCall.execute();
     }
 
-    private void buildPostPayload(HttpURLConnection httpConnection) throws IOException {
-        String paramsString = buildUrlWithParams().getQuery();
-        int bodySize = paramsString.length();
-        httpConnection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
-        httpConnection.setRequestProperty("Content-Length", bodySize + "");
-        // write to connection
-        OutputStream os = httpConnection.getOutputStream();
-        BufferedWriter writer = new BufferedWriter(
-                new OutputStreamWriter(os, "UTF-8"));
-        writer.write(paramsString);
-        writer.flush();
-        writer.close();
-        os.close();
-    }
 
-    protected URL buildUrlWithParams() throws MalformedURLException {
-        UriBuilder b = UriBuilder.fromUri(url.toString());
-        for (Map.Entry<String, String> param : params.entrySet()) {
-            b.queryParam(param.getKey(), param.getValue());
+    private Map<String, String> addContentHeaders(RequestBody postBody) {
+        headers.put("Content-Type", "application/x-www-form-urlencoded");
+        try {
+            headers.put("Content-Length", postBody.contentLength() + "");
```

Comment:
As a small point of style I'd replace this with "String.valueOf(postBody.contentLength())" to eliminate ambiguity about type coercion

Reply:
Yeah. It's a bad JAVA habit I picked up from JS